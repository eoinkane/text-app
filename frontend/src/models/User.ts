import axios, { AxiosResponse } from "axios";

interface UserProps {
  firstName?: string;
  lastName?: string;
  id?: number;
}

type valueof<T> = T[keyof T];

export default class User {
  constructor(private data: UserProps) {}

  get<K extends keyof UserProps>(propName: K): valueof<UserProps> {
    return this.data[propName];
  }

  set(update: UserProps): void {
    Object.assign(this.data, update);
  }
  fetch(): void {
    // axios
    //   .get(`http://localhost:3000/users/${this.get("id")}`)
    //   .then((response: AxiosResponse): void => {
    //     this.set(response.data);
    //   });
  }

  save(): void {
    // const id = this.get("id");
    // if (id) {
    //   axios.put(`http://localhost:3000/users/${id}`, this.data);
    // } else {
    //   axios.post(`http://localhost:3000/users`, this.data);
    // }
  }
}
