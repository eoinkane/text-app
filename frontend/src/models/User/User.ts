import axios, { AxiosResponse, AxiosError } from "axios";

interface UserProps {
  username: string;
  firstName: string;
  lastName: string;
  id: number;
}

type valueof<T> = T[keyof T];

export class CreateError extends Error {}

export type CreatePayload = {
  username: string;
  password: string;
  firstName: string;
  lastName: string;
};

export default class User {
  private data: UserProps | undefined;
  constructor() {}

  get<K extends keyof UserProps>(propName: K): valueof<UserProps> {
    if (!this.data) {
      throw new Error("User is undefined");
    }
    return this.data[propName];
  }

  private set(update: UserProps): void {
    this.data = !this.data ? update : Object.assign(this.data, update);
  }

  init(data: UserProps): void {
    this.set(data);
  }

  static async create(payload: CreatePayload): Promise<User> {
    let data: AxiosResponse | AxiosError;
    try {
      const response = await axios.post(
        "http://127.0.0.1:5000/api/users",
        {
          firstName: payload.firstName,
          lastName: payload.lastName,
        },
        {
          auth: {
            username: payload.username,
            password: payload.password,
          },
        }
      );
      data = response;

      const newUser = new User();

      newUser.init(data.data);
      return newUser;
    } catch (err) {
      console.log(err);
      throw new CreateError("Error creating new User");
    }
  }

  // save(): void {
  //   const id = this.get("id");
  //   if (id) {
  //     axios.put(`http://localhost:3000/users/${id}`, this.data);
  //   } else {
  //     axios.post(`http://localhost:3000/users`, this.data);
  //   }
  // }
}
