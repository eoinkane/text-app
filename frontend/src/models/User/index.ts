import type { CreatePayload } from "./User";
export { default } from "./User";
export { CreateError } from "./User";
export type { CreatePayload };
