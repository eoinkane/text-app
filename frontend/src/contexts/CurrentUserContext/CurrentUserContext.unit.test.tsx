import React from "react";
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import axios from "axios";
import MockAdapter from "axios-mock-adapter";
import User from "../../models/User";
import { CurrentUserProvider, initialCurrentUser } from "./index";
import {
  TestHelperParent,
  TestHelperParentWithInit,
  TestHelperChild2,
  TestHelperChild3,
  TestHelperChild4,
} from "./__mocks__/index";

export const passCurrentUserFn = jest.fn();

const mock = new MockAdapter(axios);

describe("Loading Context", () => {
  it("Renders children", () => {
    const currentUserContext = mount(
      <CurrentUserProvider>
        <TestHelperParent />
      </CurrentUserProvider>
    );
    expect(currentUserContext).toMatchSnapshot();

    expect(
      currentUserContext.find(CurrentUserProvider).children().first().type()
    ).toBe(TestHelperParent);
  });

  it("Current User is initialCurrentUser", () => {
    const currentUserContext = mount(
      <CurrentUserProvider>
        <TestHelperParent />
      </CurrentUserProvider>
    );

    expect(
      currentUserContext
        .find(CurrentUserProvider)
        .find(TestHelperChild2)
        .prop("testProp")
    ).toBe(initialCurrentUser);
  });

  it("Current User is available through `useCurrentUser` hook", () => {
    const currentUser = new User();
    currentUser.init({
      username: "test_user",
      id: 0,
      firstName: "test",
      lastName: "user",
    });
    const currentUserContext = mount(
      <CurrentUserProvider initialCurrentUserProp={currentUser}>
        <TestHelperParent />
      </CurrentUserProvider>
    );
    expect(
      currentUserContext
        .find(CurrentUserProvider)
        .find(TestHelperChild2)
        .prop("testProp")
    ).toBe(currentUser);
  });

  it("Current User can be set through `initUser` ", (done) => {
    mock.onPost("http://127.0.0.1:5000/api/users").reply(201, {
      id: 0,
      username: "test_user",
      firstName: "test",
      lastName: "user",
    });
    const currentUserContext = mount(
      <CurrentUserProvider>
        <TestHelperParentWithInit />
      </CurrentUserProvider>
    );
    act(() => {
      currentUserContext
        .find(CurrentUserProvider)
        .find(TestHelperChild3)
        .find("button")
        .simulate("click");
    });

    act(() => {
      setTimeout(() => {
        act(() => {
          currentUserContext
            .find(CurrentUserProvider)
            .find(TestHelperChild4)
            .find("button")
            .simulate("click");
        });

        act(() => {
          expect(passCurrentUserFn).toBeCalledWith({
            data: {
              firstName: "test",
              id: 0,
              lastName: "user",
              username: "test_user",
            },
          });
        });

        done();
      }, 500);
    });
  });
});
