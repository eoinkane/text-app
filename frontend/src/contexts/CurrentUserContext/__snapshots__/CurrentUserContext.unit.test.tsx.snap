// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Loading Context Renders children 1`] = `
<CurrentUserProvider>
  <TestHelperParent>
    <div>
      <TestHelperChild1>
        <TestHelperChild2
          testProp={null}
        />
      </TestHelperChild1>
    </div>
  </TestHelperParent>
</CurrentUserProvider>
`;
